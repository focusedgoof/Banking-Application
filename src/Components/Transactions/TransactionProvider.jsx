import React,{ createContext, useState} from 'react';
import PropTypes from 'prop-types';

export const TransactionContext = createContext();
 
export const TransactionProvider = ({ children }) => {
  const [transactions, setTransactions] = useState([]);

  const addTransaction = (newTransaction) => {
    setTransactions([...transactions, newTransaction]);
  };
 
  return (
    <TransactionContext.Provider value={{ transactions, addTransaction }}>
      {children}
    </TransactionContext.Provider>
  );
};

TransactionProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default TransactionProvider;
