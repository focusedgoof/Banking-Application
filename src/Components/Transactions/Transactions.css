.transactions {
  padding: 20px 0;
  border-radius: 5px;
  margin-bottom:auto;
  font-size:1.5rem;
  min-height:84.5vh;
}
  
.transactions h3 {
  font-size: 30px;
  margin-bottom:4%;
  margin-top:0;
}

.transactions .empty-table{
  width:100%;
  font-size: 1.2rem;
  font-weight:bold;
  color:#34b3f7;
  margin:auto;
}








