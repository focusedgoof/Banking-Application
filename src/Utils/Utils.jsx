export async function getLocalStorage(key) {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      try {
        const storedValue = localStorage.getItem(key);
        resolve(storedValue ? JSON.parse(storedValue) : null);
      } catch (error) {
        reject(error);
      }
    }, 0);
  });
}

export async function setLocalStorage(key, value) {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      try {
        localStorage.setItem(key, JSON.stringify(value));
        resolve();
      } catch (error) {
        reject(error);
      }
    }, 0);
  });
}
