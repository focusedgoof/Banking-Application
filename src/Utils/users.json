[
  {
    "Username": "Akshat Arya",
    "Password": "Passw0rd",
    "AccountNumber": 1234567890123456,
    "Balance": 6707,
    "PAN": "ABCD1234E",
    "Branch": "Branch A",
    "IFSCCode": "IFSC001"
  },
  {
    "Username": "Rahul Sharma",
    "Password": "Secur1ty",
    "AccountNumber": 9876543210987654,
    "Balance": 5635,
    "PAN": "WXYZ5678F",
    "Branch": "Branch B",
    "IFSCCode": "IFSC002"
  },
  {
    "Username": "Umang Mittal",
    "Password": "Log1n123",
    "AccountNumber": 4567890123456789,
    "Balance": 60375,
    "PAN": "PQRS9012G",
    "Branch": "Branch C",
    "IFSCCode": "IFSC003"
  },
  {
    "Username": "Priya Gupta",
    "Password": "NewPassw0rd",
    "AccountNumber": 7890123456789012,
    "Balance": 10000,
    "PAN": "EFGH5678H",
    "Branch": "Mumbai Central",
    "IFSCCode": "IFSC004"
  }
]
  