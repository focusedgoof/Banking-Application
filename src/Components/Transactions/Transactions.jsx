import React, { useState, useEffect } from "react";
import { getLocalStorage } from "../../Utils/Utils";
import CommonTable from "../Common/Table/CommonTable";
import "./Transactions.css";

const Transactions = () => {
  const [userTransactions, setUserTransactions] = useState([]);
  const [filteredTransactions, setFilteredTransactions] = useState([]);
  const [userAccountNumber, setUserAccountNumber] = useState("");
  const [searchText, setSearchText] = useState("");
  const [page, setPage] = useState(0);

  useEffect(() => {
    const fetchData = async () => {
      try {
        const loggedInUserData = await getLocalStorage("userData");
        if (loggedInUserData && loggedInUserData["AccountNumber"]) {
          const accountNumber = loggedInUserData["AccountNumber"];
          setUserAccountNumber(accountNumber);

          const storedTransactions =
            (await getLocalStorage("transactions")) || [];

          const userSpecificTransactions = storedTransactions
            .filter((transaction) => transaction.userAccountNumber === accountNumber)
            .map((transaction) => {
              return {
                "Account Number": transaction.beneficiaryAccountNumber.toString(),
                Date: transaction.date.toString(),
                Narration: transaction.narration.toString(),
                Debit: (transaction.debit !== 0 ? `Rs. ${transaction.debit.toString()}` : transaction.debit).toString(),
                Credit: (transaction.credit !== 0 ? `Rs. ${transaction.credit.toString()}` : transaction.credit).toString(),
                Balance: `Rs. ${transaction.balance.toString()}`,
              };
            });

          setUserTransactions(userSpecificTransactions);
          setFilteredTransactions(userSpecificTransactions);
        }
      } 
      catch (error) {
        console.error("Error fetching user data from local storage:", error);
      }
    };

    fetchData();
  }, []);

  const handleSearch = (value) => {
    setSearchText(value);
    setPage(0);

    const filtered = userTransactions.filter((transaction) => {
      const values = Object.values(transaction).join(" ").toLowerCase();
      return values.includes(value.toLowerCase());
    });

    setFilteredTransactions(filtered);
  };

  const columnHeaders = [
    { key: "Account Number", label: "Account number" },
    { key: "Date", label: "Date" },
    { key: "Narration", label: "Narration" },
    { key: "Debit", label: "Debit" },
    { key: "Credit", label: "Credit" },
    { key: "Balance", label: "Balance" },
  ];

  return (
    <section className="transactions">
      <h3>Transactions for Account Number: {userAccountNumber}</h3>      
      {userTransactions.length > 0 ? (
        <div className="common-table">
          <CommonTable
            columns={columnHeaders}
            data={filteredTransactions}
            searchText={searchText}
            handleSearch={handleSearch}
            page={page}
          />
        </div>
      ) : (
        <p className="empty-table">No existing Transaction.</p>
      )}
    </section>
  );
};

export default Transactions;
