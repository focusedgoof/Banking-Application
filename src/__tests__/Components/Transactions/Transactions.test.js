import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Transactions from '../../../Components/Transactions/Transactions';

jest.mock('../../../Utils/Utils', () => ({
  getLocalStorage: jest.fn(),
}));

describe('Transactions Component Snapshot Test', () => {
  test('renders Transactions component correctly with existing transactions', () => {
    const existingTransactions = [
      {
        userAccountNumber: '1234567890',
        beneficiaryAccountNumber: '0987654321',
        date: '2024-01-31',
        narration: 'Payment',
        debit: 100.0,
        credit: 0.0,
        balance: 5000.0,
      },
    ];

    const getLocalStorageMock = jest.fn();
    getLocalStorageMock.mockReturnValueOnce(JSON.stringify(existingTransactions));
    require('../../../Utils/Utils').getLocalStorage = getLocalStorageMock;

    const { asFragment } = render(<Transactions />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders Transactions component correctly with no existing transactions', () => {
    const getLocalStorageMock = jest.fn();
    getLocalStorageMock.mockReturnValueOnce(JSON.stringify([]));
    require('../../../Utils/Utils').getLocalStorage = getLocalStorageMock;

    const { asFragment } = render(<Transactions />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders Transactions component correctly', () => {
    const getLocalStorageMock = jest.fn();
    getLocalStorageMock.mockReturnValueOnce(JSON.stringify([]));
    require('../../../Utils/Utils').getLocalStorage = getLocalStorageMock;

    const { asFragment } = render(<Transactions />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders Transactions component correctly with one existing transaction', () => {
    const existingTransactions = [
      {
        userAccountNumber: '1234567890',
        beneficiaryAccountNumber: '0987654321',
        date: '2024-01-31',
        narration: 'Payment',
        debit: 100.0,
        credit: 0.0,
        balance: 5000.0,
      },
    ];

    const getLocalStorageMock = jest.fn();
    getLocalStorageMock.mockReturnValueOnce(JSON.stringify(existingTransactions));
    require('../../../Utils/Utils').getLocalStorage = getLocalStorageMock;

    const { asFragment } = render(<Transactions />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders Transactions component correctly with multiple existing transactions', () => {
    const existingTransactions = [
      {
        userAccountNumber: '1234567890',
        beneficiaryAccountNumber: '0987654321',
        date: '2024-01-31',
        narration: 'Payment',
        debit: 100.0,
        credit: 0.0,
        balance: 5000.0,
      },
      {
        userAccountNumber: '1234567890',
        beneficiaryAccountNumber: '1231231231',
        date: '2024-02-01',
        narration: 'Salary',
        debit: 0.0,
        credit: 2000.0,
        balance: 7000.0,
      },
    ];

    const getLocalStorageMock = jest.fn();
    getLocalStorageMock.mockReturnValueOnce(JSON.stringify(existingTransactions));
    require('../../../Utils/Utils').getLocalStorage = getLocalStorageMock;

    const { asFragment } = render(<Transactions />);
    expect(asFragment()).toMatchSnapshot();
  });
});
