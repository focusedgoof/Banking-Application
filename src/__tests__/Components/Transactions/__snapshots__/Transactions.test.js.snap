// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transactions Component Snapshot Test renders Transactions component correctly 1`] = `
<DocumentFragment>
  <section
    class="transactions"
  >
    <h3>
      Transactions for Account Number: 
    </h3>
    <p
      class="empty-table"
    >
      No existing Transaction.
    </p>
  </section>
</DocumentFragment>
`;

exports[`Transactions Component Snapshot Test renders Transactions component correctly with existing transactions 1`] = `
<DocumentFragment>
  <section
    class="transactions"
  >
    <h3>
      Transactions for Account Number: 
    </h3>
    <p
      class="empty-table"
    >
      No existing Transaction.
    </p>
  </section>
</DocumentFragment>
`;

exports[`Transactions Component Snapshot Test renders Transactions component correctly with multiple existing transactions 1`] = `
<DocumentFragment>
  <section
    class="transactions"
  >
    <h3>
      Transactions for Account Number: 
    </h3>
    <p
      class="empty-table"
    >
      No existing Transaction.
    </p>
  </section>
</DocumentFragment>
`;

exports[`Transactions Component Snapshot Test renders Transactions component correctly with no existing transactions 1`] = `
<DocumentFragment>
  <section
    class="transactions"
  >
    <h3>
      Transactions for Account Number: 
    </h3>
    <p
      class="empty-table"
    >
      No existing Transaction.
    </p>
  </section>
</DocumentFragment>
`;

exports[`Transactions Component Snapshot Test renders Transactions component correctly with one existing transaction 1`] = `
<DocumentFragment>
  <section
    class="transactions"
  >
    <h3>
      Transactions for Account Number: 
    </h3>
    <p
      class="empty-table"
    >
      No existing Transaction.
    </p>
  </section>
</DocumentFragment>
`;
